src/resource/package_builder.rs

```diff
@@ -0,0 +1,423 @@
+            references: HashMap::new(),
+            blob: PackageResourceBlob::FromMemory(data),
+        })
+    }
+
+    /// Adds a reference to the resource.
+    ///
+    /// This specifies that this resource depends on / references another resource.
+    /// Adding a reference that was previously added will overwrite its flags.
+    ///
+    /// # Arguments
+    /// * `rrid` - The resource ID of the reference.
+    /// * `flags` - The flags of the reference.
+    pub fn with_reference(&mut self, rrid: RuntimeResourceID, flags: ResourceReferenceFlagsV2) -> &mut Self {
+        self.references.insert(rrid, flags);
+        self
+    }
+
+    /// Sets the memory requirements of the resource.
+    ///
+    /// # Arguments
+    /// * `system_memory_requirement` - The system memory requirement of the resource.
+    /// * `video_memory_requirement` - The video memory requirement of the resource.
+    pub fn with_memory_requirements(&mut self, system_memory_requirement: u32, video_memory_requirement: u32) -> &mut Self {
+        self.system_memory_requirement = system_memory_requirement;
+        self.video_memory_requirement = video_memory_requirement;
+        self
+    }
+}
+
+/// The version of the package.
+///
+/// `RPKGv1` is the original version of the package format used in Hitman 2016 and Hitman 2.
+/// `RPKGv2` is the updated version of the package format used in Hitman 3.
+pub enum PackageVersion {
+    RPKGv1,
+    RPKGv2,
+}
+
+/// A builder for creating a ResourcePackage.
+pub struct PackageBuilder {
+    chunk_id: u8,
+    chunk_type: ChunkType,
+    patch_id: u8,
+    resources: HashMap<RuntimeResourceID, PackageResourceBuilder>,
+    unneeded_resources: HashSet<RuntimeResourceID>,
+}
+
+#[derive(Debug, Error)]
+pub enum PackageBuilderError {
+    #[error("Error writing the file: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Error serializing the package: {0}")]
+    SerializationError(#[from] binrw::Error),
+
+    #[error("No resources added to the package")]
+    NoResources,
+
+    #[error("Unneeded resources are only supported when building a patch package")]
+    UnneededResourcesNotSupported,
+
+    #[error("Building patch but no patch ID was provided")]
+    NoPatchId,
+
+    #[error("Too many resources in the package")]
+    TooManyResources,
+
+    #[error("A resource has too many references")]
+    TooManyReferences,
+
+    #[error("Resource type is not valid")]
+    InvalidResourceType,
+}
+
+struct OffsetTableResult {
+    offset_table_size: u32,
+    resource_entry_offsets: HashMap<RuntimeResourceID, u64>,
+}
+
+struct MetadataTableResult {
+    metadata_table_size: u32,
+    last_offset: u64,
+}
+
+impl PackageBuilder {
+    /// Creates a new package builder.
+    ///
+    /// # Arguments
+    /// * `chunk_id` - The chunk ID of the package.
+    /// * `chunk_type` - The chunk type of the package.
+    pub fn new(chunk_id: u8, chunk_type: ChunkType) -> Self {
```

Comment:
How does PartitionId relate to chunk_id and chunk_type (or any other fields)?

Reply:
[PartitionId](https://github.com/dafitius/rpkg-rs/blob/main/src/resource/pdefs/mod.rs#L58) combines the chunk_id and chunk_type into a single type. The different game versions make it difficult to work with the partition names. because chunkX, chunkXlangXX, dlcX and dlcXlangXX all need to be parsed to different things.

As a bonus: using PartitionId also allows the user to write "chunk1", instead of ```(ChunkType::Standard, 1);```